New query condition support to match all strings that starts with some other given strings

eduardbosch opened the pull request:
Hi to all! 🎉 

This changes gives support to this contributions:
https://github.com/parse-community/Parse-SDK-Android/pull/673
and
https://github.com/parse-community/Parse-SDK-JS/pull/437

It allows to send regular expressions with $all query.

**Please, read this link to know why I need this changes in the server:**
### https://github.com/parse-community/Parse-SDK-Android/pull/673

Thanks a lot 😀

flovilmart commented:
Thanks for the PR, this looks great but could you add proper edge to edge test with `it_only_db('mongo')` so we don't run it on Postgres. @vitaly-t do you think this could work on PG as well?

vitaly-t commented:
@flovilmart I can't really say. I haven't actually used parse-server for anything yet. Just doing support as far as `pg-promise` is concerned.

flovilmart commented on spec/MongoTransform.spec.js:
```diff
@@ -320,6 +320,9 @@ describe('parseObjectToMongoObjectForCreate', () => {
 
+});
+
+describe_only_db('mongo')('parseObjectToMongoObjectForCreate', () => {
```
I was thinking about an edge to edge test case, with proper data (create multiple objects and run the query). Do you think that's possible?

eduardbosch replied:
I was writing you to ask if this was what you where asking for. I don't know how to make this tests. Is there any edge to edge test case in the repository? Sorry, I'm not really familiar with Jasmine.

If you tell me how to make this tests or where to take an idea, I'll take a look for sure.

Do you think I should rollback this change??

flovilmart replied:
> Do you think I should rollback this change??

You can roll it back, but that doesn't really change anything.

The edge to edge test would look like:

```
it('should match all strings that starts with string', (done) => {
  let object = new Parse.Object('Object');
  object.set('strings', ['the', 'brown', 'fox', 'jumps']);
  let object2 = new Parse.Object('Object');
  object2.set('strings', ['over', 'the', 'lazy', 'dog']);
  Parse.Object.saveAll([object, object2]).then(() => {
   return request.get({
      url: Parse.serverURL + 'classes/Object',
      json: true,
      query: ... // build the JSON query
   });
  }).then((results) => {
     expect(....) // TODO: validate the results
  });
}); 

```

eduardbosch replied:
thanks @flovilmart 

I'll make this tests when I have some time!

eduardbosch commented:
I've added the tests with _containsAllStartingWith_ function. To simplify the tests, I updated the parse dependency to the one I've implemented the _containsAllStartingWith_ for _Parse JS SDK_.

Sorry for the delay, but I'm so busy this days.

Let me know the next steps ☺️

flovilmart commented:
@eduardbosch any chance you'd get it to work with postgres? I know that regexp support is not 💯  in PG, so perhaps you should just replace `it` by `it_exclude_dbs(['postgres'])`

dplewis commented:
@eduardbosch I have this working for postgres. You get find a solution on my branch.

https://github.com/dplewis/parse-server/commit/bc449e9333f2d48e6a6530a0b3d6bf71dad72430

You may have to add some validation on tests for postgres transform.

eduardbosch commented:
@dplewis Thanks for the changes. I've added to the PR.

@flovilmart Do you know why Travis is not checking the integration? I'm new to Travis, so I'm not really sure what's happening. I think postgres tests should work with @dplewis changes now.

In the other hand, I would like to know how to add tests to check progress for regex transform.

Thanks to all and sorry for the delay. I'll like to fix all this 😛

dplewis commented:
@eduardbosch Can you make a test for REST similar to ParseQuery.spec
https://github.com/parse-community/parse-server/blob/master/spec/ParseQuery.spec.js#L2914

The following should throw an error. Every index of $all should check for $regex
```
where: {
  strings: {
    $all: [
      {$regex: ...},
      {$regex: ...},
      {$unknown: ...}
    ]
  }
}
```

eduardbosch commented:
@dplewis Now parse server checks that all values in $all must be of type {$regex: '\^\\Qstring\\E'} or none.

dplewis commented on package.json:
```diff
@@ -29,7 +29,7 @@
     "mongodb": "2.2.26",
     "multer": "1.3.0",
-    "parse": "1.9.2",
+    "parse": "eduardbosch/Parse-SDK-JS#contains-all-starting-with",
```
Since you are using REST the JavaScript branch isn't needed. Changing it back will remove the conflict also.

dplewis commented on spec/ParseQuery.spec.js:
```diff
@@ -343,6 +343,92 @@ describe('Parse.Query testing', () => {
+        .then(function (results) {
+          equal(results.results.length, 1);
+
+          return require('request-promise').get({
```
Move this to a separate test for errors

dplewis commented on spec/ParseQuery.spec.js:
```diff
@@ -343,6 +343,92 @@ describe('Parse.Query testing', () => {
+      equal(objectList.length, results.length);
+
+      new Parse.Query('Object')
+        .containsAllStartingWith('strings', ['the', 'fox', 'lazy'])
```
REST covers this you can remove the .containsAllStartingWith

dplewis commented on src/Controllers/DatabaseController.js:
```diff
@@ -44,6 +44,30 @@ const transformObjectACL = ({ ACL, ...result }) => {
 
 const specialQuerykeys = ['$and', '$or', '_rperm', '_wperm', '_perishable_token', '_email_verify_token', '_email_verify_token_expires_at', '_account_lockout_expires_at', '_failed_login_count'];
 
+const isStartsWith = value => {
```
Can you move the checks from the controller to the adapters?

dplewis commented on src/Adapters/Storage/Postgres/PostgresStorageAdapter.js:
```diff
@@ -307,7 +307,15 @@ const buildWhereClause = ({ schema, query, index }) => {
 
     if (Array.isArray(fieldValue.$all) && isArrayField) {
-      patterns.push(`array_contains_all($${index}:name, $${index + 1}::jsonb)`);
+      if (fieldValue.$all[0].$regex) {
```
Here I just checked the first index but you can loop through and check all fields instead of checking on the controller

dplewis replied:
The same can be done for mongo

eduardbosch replied:
It's alredy done for Mongo too 🙂

eduardbosch commented:
@flovilmart @dplewis Is there any easy way to check Postgres tests without pushing changes?

dplewis commented:
https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md

eduardbosch commented:
@dplewis Thanks a lot, I haven't seen the postgres configuration part in the CONTRIBUTING.md, I'm sorry. Thanks!

eduardbosch commented:
I think now it's working like expected. Thanks a lot for your support!

Let me know if there something missing 😃

dplewis commented:
@flovilmart Can you look this over?

flovilmart commented:
I’ll have a look!

dplewis commented:
@eduardbosch ping?

eduardbosch commented:
@dplewis @flovilmart I've merged the latest changes to pass integration checks.

I think I cannot do anything else.

Is something missing?

dplewis commented on src/Adapters/Storage/Postgres/PostgresStorageAdapter.js:
```diff
@@ -1775,6 +1789,36 @@ function processRegexPattern(s) {
+
+function isAllValuesRegexOrNone(values) {
+  if (values.length == 0) {
+    return true;
```
Just looking over the code coverage. I don't see a test case for None / Empty {}

dplewis commented on src/Adapters/Storage/Postgres/PostgresStorageAdapter.js:
```diff
@@ -1775,6 +1789,36 @@ function processRegexPattern(s) {
+  }
+
+  var startsWith = isStartsWithRegex(values[0].$regex);
+  for (var i = 1, length = values.length; i < length; ++i) {
```
Interesting looking for-loop you have here. 

`for (let i = 0; i < values.length; i += 1) {}`

dplewis commented on src/Adapters/Storage/Mongo/MongoTransform.js:
```diff
@@ -123,6 +123,30 @@ const transformKeyValueForUpdate = (className, restKey, restValue, parseFormatSc
+  }
+
+  var startsWith = isStartsWithRegex(values[0]);
+  for (var i = 1, length = values.length; i < length; ++i) {
```
Same here

eduardbosch commented:
@dplewis all fixed now

Do you think I should add more tests?

dplewis commented:
@eduardbosch Still need test cases to ensure your changes function properly. 

Make separate test cases for

Empty All
```
$all: []
```
Empty Regex
```
$all: [
  { }
]
```
One Regex
```
$all: [
  {$regex: ... }
]
```
One Unknown Regex
```
$all: [
  {$unknown: ... }
]
```

I think we should be good after.

eduardbosch commented:
@dplewis tests added

I've added some tests to _spec/MongoTransform.spec.js_ also.

I had to setup different tests for Postgres and Mongo as _containsAll_ works differently in both databases.

Maybe would be better to try to unify the same functionality for both databases?

dplewis commented on spec/ParseQuery.spec.js:
```diff
@@ -497,6 +497,345 @@ describe('Parse.Query testing', () => {
+        }
+      });
+    })
+      .then(function () {
```
@eduardbosch I figured they would return different values for both databases. Some of your tests are missing `.then()` return values. If you add the return values for PG I can look at the script I wrote to see if I can get similar functionality to mongo

eduardbosch replied:
@dplewis I don't understand. Some miss `.then()` return values because they throw an error, so that function is never called.

dplewis replied:
Sorry missed that. Looking at your test, it should be easy to make PG throw an error instead of returning zero results for invalid $all.

dplewis commented on spec/MongoTransform.spec.js:
```diff
@@ -331,6 +331,41 @@ describe('parseObjectToMongoObjectForCreate', () => {
+    done();
+  });
+
+  it('all values in $all must be $regex (start with string) or non $regex (start with string)', (done) => {
```
Thanks for adding the unit tests here. I know you are checking for isArray. I'm curious to find out what would happen if I put

$all: false
$all: 1234
$all: undefined
$all: null
$all: {}

eduardbosch commented:
Sorry for the long delay!

I've unified all functionality from MongoDB and PostgreSQL except when the $all array is empty.

In MongoDB returns an empty result but in PostgreSQL returns all results.

Maybe `contains-all-regex.sql` should be changed to return empty results on empty $all.

dplewis commented:
@eduardbosch @flovilmart Sorry for the late reply. The empty `$all` has another issue here. 

`query.containAll('field', [])` returns all values for PG and have nothing to do with this PR.

Basically `contains-all-regex.sql` doesn't get called here because there aren't any regex present but `contains-all.sql` does.

This should be good to merge. I'll create another PR for the empty $all

flovilmart commented:
As it stands because there’s no edge to edge tests, it’s passing but adding a proper test would let it fail. Do you believe $nor and $elemMatch would be portable to PG?

dplewis commented:
@flovilmart Wrong thread 😜. I can look at it if there are test cases.

flovilmart commented:
Oops! Sorry!!

dplewis commented:
@eduardbosch Sorry this took almost a year. 🎉. We learned a lot about PG and found a bug. Thanks for the contribution.

dplewis commented:
@flovilmart How does this look?

flovilmart commented:
Thanks @eduardbosch for your commitment and patience!

eduardbosch commented:
Thanks to you @flovilmart @dplewis for your patience too!! I've learned a lot too 🎉

So happy to see this merged 😍I hope this could be a useful feature for someone.

I think we should start taking a look at the SDKs integrations now:
* https://github.com/parse-community/Parse-SDK-JS/pull/437
* https://github.com/parse-community/Parse-SDK-Android/pull/673

I'm currently porting my app to iOS, so I think I'll add those changes to the iOS SDK too in a future.
